haskell-src/Concordium/Types/ProtocolLevelTokens/CBOR.hs

```diff
@@ -1324,6 +1328,76 @@ data TokenModuleState = TokenModuleState
     }
     deriving (Eq, Show)
 
+-- | Function should be already available after merging "https://github.com/Concordium/concordium-base/pull/657"
+-- | Convert CBOR.Term to a hex-encoded string
+cborTermToHex :: CBOR.Term -> Text
+cborTermToHex term =
+    let bs = CBOR.toStrictByteString $ CBOR.encodeTerm term
+    in  TextEncoding.decodeUtf8 (Base16.encode bs)
+
+-- | Function should be already available after merging "https://github.com/Concordium/concordium-base/pull/657"
+-- | Convert a hex-encoded string to a CBOR.Term.
+hexToCborTerm :: Text -> Either String CBOR.Term
+hexToCborTerm hexText = do
+    bs <- Base16.decode (TextEncoding.encodeUtf8 hexText)
+    decodeTerm bs
+  where
+    decodeTerm bs =
+        case CBOR.deserialiseFromBytes CBOR.decodeTerm (LBS.fromStrict bs) of
+            Left err -> Left $ "Failed to decode CBOR term: " ++ show err
+            Right ("", term) -> Right term
+            Right (remaining, _) -> Left $ "Extra bytes after decoding CBOR term: " ++ show (LBS.length remaining)
+
+instance AE.ToJSON TokenModuleState where
+    toJSON TokenModuleState{..} =
+        AE.object
+            ( [ "name" AE..= tmsName,
+                "metadata" AE..= tmsMetadata,
+                "allowList" AE..= tmsAllowList,
+                "denyList" AE..= tmsDenyList,
+                "mintable" AE..= tmsMintable,
+                "burnable" AE..= tmsBurnable
+              ]
+                ++ maybe
+                    []
+                    ( \pairs ->
+                        ["_additional" AE..= AE.object pairs]
+                    )
+                    additional
+            )
+      where
+        additional
+            | null tmsAdditional = Nothing
+            | otherwise =
+                Just
+                    [ AE.Key.fromText k AE..= cborTermToHex v
+                      | (k, v) <- Map.toList tmsAdditional
+                    ]
+
+instance AE.FromJSON TokenModuleState where
+    parseJSON = AE.withObject "TokenModuleState" $ \v -> do
+        tmsName <- v AE..: "name"
+        tmsMetadata <- v AE..: "metadata"
+        tmsAllowList <- v AE..: "allowList"
+        tmsDenyList <- v AE..: "denyList"
+        tmsMintable <- v AE..: "mintable"
+        tmsBurnable <- v AE..: "burnable"
+        tmsAdditionalTemp <- v AE..:? "_additional" .!= Map.empty
```

Comment:
I think you can avoid the `tmsAdditionalTemp` and directly parse `tmsAdditional`.